flags_test.go

```diff
@@ -357,3 +361,117 @@ func (t *FlagsTest) TestValidateFlagsForValidSequentialReadSizeAndHTTP2ClientPro
+			statCacheTTL:             DefaultStatOrTypeCacheTTL,
+			typeCacheTTL:             DefaultStatOrTypeCacheTTL,
+			ttlInSeconds:             0,
+			expectedMetadataCacheTTL: 0,
+		},
+		{
+			// scenario where user sets only metadata-cache:ttl-secs and sets it to a positive value
+			statCacheTTL:             DefaultStatOrTypeCacheTTL,
+			typeCacheTTL:             DefaultStatOrTypeCacheTTL,
+			ttlInSeconds:             30,
+			expectedMetadataCacheTTL: 30 * time.Second,
+		},
+		{
+			// scenario where user sets only metadata-cache:ttl-secs and sets it to its highest supported value
+			statCacheTTL: DefaultStatOrTypeCacheTTL,
+			typeCacheTTL: DefaultStatOrTypeCacheTTL,
+			ttlInSeconds: config.MaxSupportedTtlInSeconds,
+
+			// math.MaxInt64 is not a multiple of 1e9, so time.Duration(math.MaxInt64) is not
+			// an exact multiple of seconds. Since ttlInSeconds will always set time in
+			// multiple of seconds, we also need to round time.Duration(math.MaxInt64) to
+			// multiple of seconds, for comparison.
+			expectedMetadataCacheTTL: time.Duration((math.MaxInt64 / int64(time.Second)) * int64(time.Second)),
+		},
+		{
+			// scenario where user sets both the old flags and the metadata-cache:ttl-secs. Here ttl-secs overrides both flags. case 1.
+			statCacheTTL:             5 * time.Minute,
+			typeCacheTTL:             time.Hour,
+			ttlInSeconds:             10800,
+			expectedMetadataCacheTTL: 3 * time.Hour,
+		},
+		{
+			// scenario where user sets both the old flags and the metadata-cache:ttl-secs. Here ttl-secs overrides both flags. case 2.
+			statCacheTTL:             5 * time.Minute,
+			typeCacheTTL:             time.Hour,
+			ttlInSeconds:             1800,
+			expectedMetadataCacheTTL: 30 * time.Minute,
+		},
+		{
+			// scenario where user sets both the old flags and the metadata-cache:ttl-secs. Here ttl-secs overrides both flags. case 3.
+			statCacheTTL:             5 * time.Minute,
+			typeCacheTTL:             time.Hour,
+			ttlInSeconds:             0,
+			expectedMetadataCacheTTL: 0,
+		},
+		{
+			// scenario where user sets both the old flags and the metadata-cache:ttl-secs. Here ttl-secs overrides both flags. case 4.
+			statCacheTTL:             5 * time.Minute,
+			typeCacheTTL:             time.Hour,
+			ttlInSeconds:             -1,
+			expectedMetadataCacheTTL: time.Duration(math.MaxInt64),
+		},
+		{
+			// old-scenario where user sets only stat/type-cache-ttl flag(s), and not metadata-cache:ttl-secs. Case 1.
+			statCacheTTL:             0,
+			typeCacheTTL:             0,
+			ttlInSeconds:             config.TtlInSecsUnsetSentinel,
+			expectedMetadataCacheTTL: 0,
+		},
+		{
+			// old-scenario where user sets only stat/type-cache-ttl flag(s), and not metadata-cache:ttl-secs. Case 2. Stat-cache enabled, but not type-cache.
+			statCacheTTL:             time.Hour,
+			typeCacheTTL:             0,
+			ttlInSeconds:             config.TtlInSecsUnsetSentinel,
+			expectedMetadataCacheTTL: 0,
```

Comment:
This is according to the agreed-upon config design i.e. `ttl = min(stat-cache-ttl, type-cache-ttl)`. But this looks wrong to me. Here (and in the next test case), user is enabling one out of stat/type cache and is disabling the other, we end up disabling both (by setting `metadataCacheTTL=0`. This looks like bad behaviour to me.

Reply:
There is no natural solution to this if we merge the TTLs into one. We could technically support it by internally maintaining two different ttls for the special case that one is enabled and the other is disabled, but it'll give rise to a discontinuity in the behavior. I'll keep this as an open question for myself for later.